File: prologapp1/game.pl
%% 推箱子核心逻辑（Prolog 教学，SWI-Prolog）

:- module(game, [
    from_rows/2,
    try_move/4,
    undo/2,
    render_ascii/2,
    state_moves/2,
    state_won/2
]).

%% state(Walls, Goals, Boxes, PX-PY, Moves, Won, Width, Height, Hist)
%% Walls/Goals/Boxes: list of X-Y

cell_key(X, Y, X-Y).

from_rows(Rows, state(Walls, Goals, Boxes, PX-PY, 0, false, W, H, [])) :-
    parse_rows(Rows, 0, [], [], [], 0-0, 0, 0, Walls, Goals, Boxes, PX-PY, MaxX, MaxY),
    W is MaxX + 1,
    H is MaxY + 1.

parse_rows([], _, Walls, Goals, Boxes, Pos, MaxX, MaxY, Walls, Goals, Boxes, Pos, MaxX, MaxY).
parse_rows([Row|Rest], Y, Walls0, Goals0, Boxes0, Pos0, MaxX0, MaxY0,
           Walls, Goals, Boxes, Pos, MaxX, MaxY) :-
    atom_chars(Row, Chars),
    parse_row(Chars, 0, Y, Walls0, Goals0, Boxes0, Pos0, MaxX0,
              Walls1, Goals1, Boxes1, Pos1, MaxX1),
    MaxY1 is max(MaxY0, Y),
    Y1 is Y + 1,
    parse_rows(Rest, Y1, Walls1, Goals1, Boxes1, Pos1, MaxX1, MaxY1,
               Walls, Goals, Boxes, Pos, MaxX, MaxY).

parse_row([], _, _, W, G, B, P, MX, W, G, B, P, MX).
parse_row([Ch|Rest], X, Y, Walls0, Goals0, Boxes0, Pos0, MaxX0,
          Walls, Goals, Boxes, Pos, MaxX) :-
    MaxX1 is max(MaxX0, X),
    apply_cell(Ch, X, Y, Walls0, Goals0, Boxes0, Pos0,
               Walls1, Goals1, Boxes1, Pos1),
    X1 is X + 1,
    parse_row(Rest, X1, Y, Walls1, Goals1, Boxes1, Pos1, MaxX1,
              Walls, Goals, Boxes, Pos, MaxX).

apply_cell('#', X, Y, W, G, B, P, [X-Y|W], G, B, P).
apply_cell('.', X, Y, W, G, B, P, W, [X-Y|G], B, P).
apply_cell('$', X, Y, W, G, B, P, W, G, [X-Y|B], P).
apply_cell('*', X, Y, W, G, B, P, W, [X-Y|G], [X-Y|B], P).
apply_cell('@', X, Y, W, G, B, _, W, G, B, X-Y).
apply_cell('+', X, Y, W, G, B, _, W, [X-Y|G], B, X-Y).
apply_cell(_, _, _, W, G, B, P, W, G, B, P).

member_pos(X-Y, List) :- member(X-Y, List).

check_win(Boxes, Goals) :-
    forall(member(Box, Boxes), member(Box, Goals)).

state_moves(state(_,_,_,_,M,_,_,_,_), M).
state_won(state(_,_,_,_,_,Won,_,_,_), Won).

try_move(state(_,_,_,_,_,true,_,_,_), _, _, _) :- !, fail.
try_move(state(Walls, Goals, Boxes, PX-PY, Moves, _, W, H, Hist), DX, DY,
         state(Walls, Goals, Boxes2, NX-NY, Moves2, Won2, W, H, Hist2)) :-
    NX is PX + DX,
    NY is PY + DY,
    \+ member_pos(NX-NY, Walls),
    ( member_pos(NX-NY, Boxes) ->
        BX is NX + DX,
        BY is NY + DY,
        \+ member_pos(BX-BY, Walls),
        \+ member_pos(BX-BY, Boxes),
        select(NX-NY, Boxes, Boxes1),
        Boxes2 = [BX-BY|Boxes1],
        Moves2 is Moves + 1,
        Hist2 = [hist(PX-PY, NX-NY, BX-BY)|Hist],
        ( check_win(Boxes2, Goals) -> Won2 = true ; Won2 = false )
    ;
        Boxes2 = Boxes,
        Moves2 = Moves,
        Won2 = false,
        Hist2 = [hist(PX-PY, none, none)|Hist]
    ).

undo(state(_,_,_,_,_,true,_,_,_), _) :- !, fail.
undo(state(_,_,_,_,_,_, _,_,[]), _) :- !, fail.
undo(state(Walls, Goals, Boxes, _, Moves, _, W, H, Hist),
     state(Walls, Goals, Boxes2, PX-PY, Moves2, false, W, H, Rest)) :-
    undo_loop(Hist, Boxes, Moves, Boxes2, PX-PY, Moves2, Rest).

%% 与其它语言版一致：回退到上一次推箱（中间走路一并撤销）
undo_loop([hist(P, none, none)|Rest], Boxes, Moves, Boxes2, Pos, Moves2, HistOut) :-
    !,
    ( Rest = [] ->
        Boxes2 = Boxes, Pos = P, Moves2 = Moves, HistOut = []
    ; undo_loop(Rest, Boxes, Moves, Boxes2, Pos, Moves2, HistOut)
    ).
undo_loop([hist(P, BF, BT)|Rest], Boxes, Moves, Boxes2, P, Moves2, Rest) :-
    BF \= none,
    select(BT, Boxes, Boxes1),
    Boxes2 = [BF|Boxes1],
    Moves2 is max(0, Moves - 1).

render_ascii(state(Walls, Goals, Boxes, PX-PY, _, _, Width, Height, _), Text) :-
    render_rows(0, Height, Width, Walls, Goals, Boxes, PX-PY, Lines),
    atomic_list_concat(Lines, '\n', Body),
    atom_concat(Body, '\n', Text).

render_rows(Y, Height, _, _, _, _, _, []) :- Y >= Height, !.
render_rows(Y, Height, Width, Walls, Goals, Boxes, Player, [Line|Rest]) :-
    render_row(0, Width, Y, Walls, Goals, Boxes, Player, Chars),
    atomic_list_concat(Chars, '', Line),
    Y1 is Y + 1,
    render_rows(Y1, Height, Width, Walls, Goals, Boxes, Player, Rest).

render_row(X, Width, _, _, _, _, _, []) :- X >= Width, !.
render_row(X, Width, Y, Walls, Goals, Boxes, PX-PY, [Ch|Rest]) :-
    ( X =:= PX, Y =:= PY ->
        ( member_pos(X-Y, Goals) -> Ch = '+' ; Ch = '@' )
    ; member_pos(X-Y, Boxes) ->
        ( member_pos(X-Y, Goals) -> Ch = '*' ; Ch = '$' )
    ; member_pos(X-Y, Walls) ->
        Ch = '#'
    ; member_pos(X-Y, Goals) ->
        Ch = '.'
    ;
        Ch = ' '
    ),
    X1 is X + 1,
    render_row(X1, Width, Y, Walls, Goals, Boxes, PX-PY, Rest).

File: erlangapp1/game.erl
%% 推箱子核心逻辑（Erlang 教学）
-module(game).
-export([from_rows/1, try_move/3, undo/1, render_ascii/1, moves/1, won/1]).

%% state = {Walls, Goals, Boxes, {PX,PY}, Moves, Won, W, H, Hist}
%% sets as lists of {X,Y}

key(X, Y) -> {X, Y}.

from_rows(Rows) ->
    {Walls, Goals, Boxes, Pos, MaxX, MaxY} = parse(Rows, 0, [], [], [], {0,0}, 0, 0),
    {Walls, Goals, Boxes, Pos, 0, false, MaxX + 1, MaxY + 1, []}.

parse([], _, W, G, B, P, MX, MY) -> {W, G, B, P, MX, MY};
parse([Row|Rest], Y, W, G, B, P, MX, MY) ->
    {W1, G1, B1, P1, MX1} = parse_row(Row, 0, Y, W, G, B, P, MX),
    parse(Rest, Y + 1, W1, G1, B1, P1, MX1, max(MY, Y)).

parse_row([], _, _, W, G, B, P, MX) -> {W, G, B, P, MX};
parse_row([Ch|Rest], X, Y, W, G, B, P, MX) ->
    MX1 = max(MX, X),
    {W1, G1, B1, P1} = apply_ch(Ch, X, Y, W, G, B, P),
    parse_row(Rest, X + 1, Y, W1, G1, B1, P1, MX1).

apply_ch($#, X, Y, W, G, B, P) -> {[key(X,Y)|W], G, B, P};
apply_ch($., X, Y, W, G, B, P) -> {W, [key(X,Y)|G], B, P};
apply_ch($$, X, Y, W, G, B, P) -> {W, G, [key(X,Y)|B], P};
apply_ch($*, X, Y, W, G, B, P) -> {W, [key(X,Y)|G], [key(X,Y)|B], P};
apply_ch($@, X, Y, W, G, B, _) -> {W, G, B, key(X,Y)};
apply_ch($+, X, Y, W, G, B, _) -> {W, [key(X,Y)|G], B, key(X,Y)};
apply_ch(_, _, _, W, G, B, P) -> {W, G, B, P}.

member_pos(K, L) -> lists:member(K, L).

check_win(Boxes, Goals) ->
    lists:all(fun(B) -> member_pos(B, Goals) end, Boxes).

moves({_, _, _, _, M, _, _, _, _}) -> M.
won({_, _, _, _, _, Won, _, _, _}) -> Won.

try_move(State, DX, DY) ->
    {Walls, Goals, Boxes, {PX, PY}, Moves, Won, W, H, Hist} = State,
    case Won of
        true -> State;
        false ->
            NX = PX + DX, NY = PY + DY,
            NK = key(NX, NY),
            case member_pos(NK, Walls) of
                true -> State;
                false ->
                    case member_pos(NK, Boxes) of
                        true ->
                            BX = NX + DX, BY = NY + DY, BK = key(BX, BY),
                            case member_pos(BK, Walls) orelse member_pos(BK, Boxes) of
                                true -> State;
                                false ->
                                    Boxes2 = [BK | lists:delete(NK, Boxes)],
                                    Won2 = check_win(Boxes2, Goals),
                                    {Walls, Goals, Boxes2, {NX, NY}, Moves + 1, Won2, W, H,
                                     [{{PX,PY}, NK, BK}|Hist]}
                            end;
                        false ->
                            {Walls, Goals, Boxes, {NX, NY}, Moves, false, W, H,
                             [{{PX,PY}, none, none}|Hist]}
                    end
            end
    end.

undo(State) ->
    {Walls, Goals, Boxes, _, Moves, Won, W, H, Hist} = State,
    case Won orelse Hist =:= [] of
        true -> State;
        false -> undo_loop(Hist, Boxes, Moves, Walls, Goals, W, H)
    end.

undo_loop([], Boxes, Moves, Walls, Goals, W, H) ->
    {Walls, Goals, Boxes, {0,0}, Moves, false, W, H, []};
undo_loop([{{PX,PY}, none, none}|Rest], Boxes, Moves, Walls, Goals, W, H) ->
    case Rest of
        [] -> {Walls, Goals, Boxes, {PX,PY}, Moves, false, W, H, []};
        _ -> undo_loop(Rest, Boxes, Moves, Walls, Goals, W, H)
    end;
undo_loop([{{PX,PY}, BF, BT}|Rest], Boxes, Moves, Walls, Goals, W, H) when BF =/= none ->
    Boxes2 = [BF | lists:delete(BT, Boxes)],
    {Walls, Goals, Boxes2, {PX,PY}, max(0, Moves - 1), false, W, H, Rest}.

render_ascii({Walls, Goals, Boxes, {PX, PY}, _, _, Width, Height, _}) ->
    lists:flatten([render_row(Y, Width, Walls, Goals, Boxes, PX, PY) || Y <- lists:seq(0, Height - 1)]).

render_row(Y, Width, Walls, Goals, Boxes, PX, PY) ->
    [cell(X, Y, Walls, Goals, Boxes, PX, PY) || X <- lists:seq(0, Width - 1)] ++ "\n".

cell(X, Y, Walls, Goals, Boxes, PX, PY) ->
    K = key(X, Y),
    if
        X =:= PX, Y =:= PY ->
            case member_pos(K, Goals) of true -> $+; false -> $@ end;
        true ->
            case member_pos(K, Boxes) of
                true -> case member_pos(K, Goals) of true -> $*; false -> $$ end;
                false ->
                    case member_pos(K, Walls) of
                        true -> $#;
                        false ->
                            case member_pos(K, Goals) of true -> $.; false -> $\s end
                    end
            end
    end.
